paasta_tools/automatic_rollbacks.py

```diff
@@ -110,3 +118,81 @@ def watch_for_slack_webhooks():
 if __name__ == '__main__':
     logging.basicConfig(level=logging.DEBUG)
     watch_for_slack_webhooks()
+
+
+class TransitionDefinition(TypedDict):
+    trigger: str
+    source: str
+    dest: str
+
+
+class DeploymentProcess(abc.ABC):
+    if TYPE_CHECKING:
+        # These attributes need to be defined in this `if TYPE_CHECKING` block, because if they exist at runtime then
+        # transitions will refuse to overwrite them.
+        state: str
+
+        def trigger(self, *args, **kwargs):
+            ...
+
+    def __init__(
+        self,
+    ):
+
+        self.event_loop = asyncio.get_event_loop()
+        self.finished_event = asyncio.Event(loop=self.event_loop)
+
+        self.machine = transitions.extensions.LockedMachine(
+            model=self,
+            states=list(self.states()),
+            transitions=list(self.valid_transitions()),
+            initial=self.start_state(),
+            after_state_change=self.after_state_change,
+            queued=True,
+        )
+
+        for state in list(self.states()):
+            # Call e.g. on_enter_start(self.enter_start)
```

Comment:
Why do we need this? The docs say this is automatic?

Reply:
Oh, I missed the bit where `on_enter_*` on your model gets automatically called; I only saw the part where you can pass a function/method name to `on_enter_*` on your Machine.